### DISCUSSION TITLE
How to create a VProgressLinear?

### DISCUSSION THREAD

**Discussion by jipenaflor** [Kitware/trame#274] [Q&A] [answered]
Hi! I am using Trame to run an OpenFOAM application. I would like to create a progress bar (that would update whenever an OpenFOAM process is done) using the VProgressLinear. I want to do it with a v_model but I do not know how will I call a function to update the progress bar from time to time. I want to ask how to do it and how will I be able to increment it with a specific value? Thanks.

**Comment by jourdain** [comment]
So ideally all your server stuff use async processes so you never lock it. 
Then you just need to call a function like

```python
def update_progress(delta):
     with state:
          state.progress += delta
```

Assuming you have somewhere in you UI

```python
with layout:
    with layout.toolbar:
         vuetify.VProgressLinear(
            absolute=True,
            bottom=True,
            height=4,
            v_model=("progress", 0),
         )
```

Where the max value is 100.

**Reply by jipenaflor** [reply]
Hi! Sorry it took me a while to reply. I've actually followed your recommendations in the other comment, that is to implement the app with a class, and I now find it more optimized than ever. I just currently have questions about VBtn and VProgressLinear, and for reference you can see the full code [here.]( https://github.com/jipenaflor/ventilation-simulator/blob/main/ventilation_simulator/app/core.py)

1. About the usage of VProgressLinear, I followed exactly what was stated above (without the bottom property), of putting it [in UI](https://github.com/jipenaflor/ventilation-simulator/blob/main/ventilation_simulator/app/core.py#L610-L616) and updating it through a [class method](https://github.com/jipenaflor/ventilation-simulator/blob/main/ventilation_simulator/app/core.py#L197-L199). However, when I call its [update function](https://github.com/jipenaflor/ventilation-simulator/blob/main/ventilation_simulator/app/core.py#L296-L307) after an OpenFOAM process, I don't find the change in its state variable to be reflected in the progress bar. I did try printing its value every call and I see it changing, but the progress bar does not move at all. I also tried [binding the method to the state change](https://github.com/jipenaflor/ventilation-simulator/blob/main/ventilation_simulator/app/core.py#L51) but it produced a lot of errors so I presume that's not the way to go. May I know where did I go wrong?

      Moreover, whenever I click the progress bar, its value move towards the point where I clicked, where the mouse is pointing. I do not if it is a bug or I just really used the VProgressLinear incorrectly.

2. About the VBtn, I would like to disable the [set button](https://github.com/jipenaflor/ventilation-simulator/blob/main/ventilation_simulator/app/core.py#L602-L607) while the [set function](https://github.com/jipenaflor/ventilation-simulator/blob/main/ventilation_simulator/app/core.py#L296-L307) to avoid getting it clicked multiple times, calling the set function while the OpenFOAM processes are running. May I know if that is possible? Thanks!

Also, thank you for looking forward to my undergraduate capstone project. I really appreciate the support! I am also looking forward to further development of Trame and to be able to use it again in my future projects.

**Reply by jourdain** [reply]
So the problem with your code is that you keep the server busy so it can not handle the network and let the UI refresh. A way to support it would be to add a `await asyncio.sleep(0.01)` after each `update_setProgress` call. That will force you to define your `set` function to be async. But then you should not call it directly from the button click either. You want to release the control as well by running it as a task. Something like (keep the class method just providing an example as global function for simplicity)

```python
import asyncio
from trame.app import asynchronous

async def _async_run():
        if self.inlet != self.outlet:
            if self.toSet:
                self.convert()
                self.update_setProgress(5)
                await asyncio.sleep(0.01)
                self.block()
                self.update_setProgress(15)
                await asyncio.sleep(0.01)
                self.mesh()
                self.update_setProgress(75)
                await asyncio.sleep(0.01)
                self.view_stl()
                self.update_setProgress(5)
                await asyncio.sleep(0.01)
                self.setSuccess = True
                with state:
                    state.sim_running = False


def run_sim():
     state.sim_running = True
     asynchronous.create_task(_async_run())

with layout:
    # ...
     vuetify.VBtn(
          "Set",
          click=run_sim,
          variant="tonal",
          disabled=("sim_running", False),
           classes="pa-3"
     )

**Reply by jipenaflor** [reply]
It's now working as desired. Thanks!

**Comment by jourdain** [comment, marked as answer]
BTW, I was curious and excited by your project and have a couple comments

1. When you have a `@state.change(...)`, your [decorated function](https://github.com/jipenaflor/online_ventilation_simulator/blob/main/app.py#L292) must have a `**kwargs` since the full state is passed as arg. 
2. When you react to state change, [you should not change that same state](https://github.com/jipenaflor/online_ventilation_simulator/blob/main/app.py#L320C16-L337) variable again. Basically that will trigger the same function execution with no matching if.
3. You should not put [thing in the state](https://github.com/jipenaflor/online_ventilation_simulator/blob/main/app.py#L306-L307) that are not meant for the UI. Instead use a class with internal variable or something like that. That server state is meant for sending things over the network between the Web UI and your Python server in either direction. So when you do [that](https://github.com/jipenaflor/online_ventilation_simulator/blob/main/app.py#L126) you just received the content of a file on the server side and you push it back to the client right away in another variable name. You can still stick to a single file app, but if you want some idea on how to use class rather than global module level functions, you should try the basic app from the [trame-cookiecutter](https://github.com/kitware/trame-cookiecutter).

Thanks again for exploring trame and good luck with your project as I can't wait so see more simulation app using trame!
